proposals/exception-handling/Exceptions.md

```diff
@@ -205,99 +193,166 @@ flushed, the embedder defines how to handle uncaught exceptions. Otherwise, the
+`rethrow 0` corresponds to the exception caught by `catch 3`, `rethrow 1`
+corresponds to the exception caught by `catch 2`, and `rethrow 3` corresponds
+to the exception caught by `catch 1`. In wat format, the argument for the
+`rethrow` instructions can also be written as a label, like branches. So
+`rethrow 0` in the example above can also be written as `rethrow $l3`.
+
+Note that `rethrow 2` is not allowed because it does not reference a `try`
+instruction. Rather, it references a `block` instruction, so it will result in a
+validation failure.
+
+Note that the example below is a validation failure:
+```
+try $l1
+  try $l2
+    rethrow $l2 (= rethrow 0)
+  catch
+  end
+catch
+end
+```
+The `rethrow` here references `try $l2`, but the `rethrow` is not within its
+`catch` block.
+
+### Try-delegate blocks
+
+Try blocks can also be used with the `delegate` instruction. A try-delegate
+block contains an `delegate` instruction with the following form:
+
+```
+try blocktype
+  instruction*
+delegate label
 ```
 
-The `br_on_exn` instruction checks the exception tag of an `exnref` on top of
-the stack if it matches the given exception index. If it does, it branches out
-to the label referenced by the instruction (In the binary form, the label will
-be converted to a relative depth immediate, like other branch instructions), and
-while doing that, pops the `exnref` value from the stack and instead pushes the
-exception's argument values on top of the stack. In order to use these popped
-values, the block signature of the branch target has to match the exception
-types - because it receives the exception arguments as branch operands. If the
-exception tag does not match, the `exnref` value remains on the stack. For
-example, when an `exnref` contains an exception of type (i32 i64), the target
-block signature should be (i32 i64) as well, as in the following example:
+The `delegate` instruction does not have an associated body, so try-delegate
+blocks don't have an `end` instruction at the end. The `delegate` instruction
+takes a label defined by a construct in which they are enclosed, and delegates
+exception handling to a catch/unwind block specified by the label. For example,
+consider this code:
 
 ```
-block $l (result i32 i64)
+try $l1
+  try
+    call $foo
+  delegate $l1 (= delegate 0)
+catch
   ...
-  ;; exnref $e is on the stack at this point
-  br_on_exn $l ;; branch to $l with $e's arguments
+catch_all
   ...
 end
 ```
 
-This can now be used to construct handler switches in the same way `br_table`
-is used to construct regular switch:
+If `call $foo` throws, searching for a catching block first finds `delegate`,
+and because it delegates exception handling to catching instructions associated
+with `$label0`, it will be next checked by the outer `catch` and then
+`catch_all` instructions.
+When the specified label within a `delegate`
+instruction does not correspond to a `try` instruction, it is a validation
+failure.
```

Comment:
Technically, this is an unnecessary restriction as well. We could specify it as simply being equivalent to a rethrow inside the target block, which makes it agnostic to what sort of construct it is. That also increases flexibility, because it allows to delegate to an enclosing handler that's not in lexical scope, which is more in line with the dynamically scoped nature of exception handlers.

Reply:
We discussed this at length in some of comments in #143 and #146, and it seems people there, including me, agreed that validation failure is clearer. I also would like to strongly advocate for this, because assuming exception can slip until it meets any enclosing try will make toolchain optimization, such as in Binaryen, very difficult.